control/controlclient/noise.go

```diff
@@ -295,11 +295,27 @@ func (nc *NoiseClient) getConn(ctx context.Context) (*noiseConn, error) {
 	}
 	nc.mu.Unlock()
 
-	conn, err, _ := nc.sfDial.Do(struct{}{}, nc.dial)
-	if err != nil {
-		return nil, err
+	for {
+		// We singeflight the dial to avoid making multiple connections, however
+		// that means that we can't simply cancel the dial if the context is
+		// canceled. Instead, we have to additionally check that the context
+		// which was canceled is our context and retry if our context is still
+		// valid.
+		conn, err, _ := nc.sfDial.Do(struct{}{}, func() (*noiseConn, error) {
+			return nc.dial(ctx)
+		})
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
```

Comment:
not just Canceled. Could also be deadline exceeded.

Reply:
maybe don't look at this at all and only look at ctx.Err() being non-nil?

Reply:
Redid this, PTAL